examples_tests/02.ComputeShader/main.cpp

```diff
@@ -27,11 +38,584 @@ void main()
+
+		memcpy(mappedMemoryAddress, &uboData_cpu[i], sizeof(UniformBufferObject));
+
+		device->unmapMemory(ubosMemory[i].get());
+	}
+
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		const uint32_t writeDescriptorCount = 2u;
+
+		video::IGPUDescriptorSet::SDescriptorInfo descriptorInfos[writeDescriptorCount];
+		video::IGPUDescriptorSet::SWriteDescriptorSet writeDescriptorSets[writeDescriptorCount] = {};
+
+		// image2D
+		{
+			descriptorInfos[0].image.imageLayout = asset::EIL_GENERAL;
+			descriptorInfos[0].image.sampler = nullptr;
+			descriptorInfos[0].desc = swapchainImageViews[i]; // shouldn't IGPUDescriptorSet hold a reference to the resources in its descriptors?
+
+			writeDescriptorSets[0].dstSet = descriptorSets[i].get();
+			writeDescriptorSets[0].binding = 0u;
+			writeDescriptorSets[0].arrayElement = 0u;
+			writeDescriptorSets[0].count = 1u;
+			writeDescriptorSets[0].descriptorType = asset::EDT_STORAGE_IMAGE;
+			writeDescriptorSets[0].info = &descriptorInfos[0];
+		}
+
+		// ubo
+		{
+			descriptorInfos[1].buffer.offset = 0ull;
+			descriptorInfos[1].buffer.size = sizeof(UniformBufferObject);
+			descriptorInfos[1].desc = ubos[i];
+
+			writeDescriptorSets[1].dstSet = descriptorSets[i].get();
+			writeDescriptorSets[1].binding = 1u;
+			writeDescriptorSets[1].arrayElement = 0u;
+			writeDescriptorSets[1].count = 1u;
+			writeDescriptorSets[1].descriptorType = asset::EDT_UNIFORM_BUFFER;
+			writeDescriptorSets[1].info = &descriptorInfos[1];
+		}
+
+		device->updateDescriptorSets(writeDescriptorCount, writeDescriptorSets, 0u, nullptr);
+	}
+
+	// Todo(achal): Make use of push constants for something, just to test
+	core::smart_refctd_ptr<video::IGPUPipelineLayout> pipelineLayout =
+		device->createGPUPipelineLayout(nullptr, nullptr, core::smart_refctd_ptr(dsLayout));
+
+	core::smart_refctd_ptr<video::IGPUComputePipeline> pipeline =
+		device->createGPUComputePipeline(nullptr, core::smart_refctd_ptr(pipelineLayout),
+			core::smart_refctd_ptr(specializedShader));
+
+	// Todo(achal): Consider making it greater than swapchainImageCount
+	const uint32_t FRAMES_IN_FLIGHT = 2u;
+
+	core::smart_refctd_ptr<video::IGPUSemaphore> acquireSemaphores[FRAMES_IN_FLIGHT];
+	core::smart_refctd_ptr<video::IGPUSemaphore> releaseSemaphores[FRAMES_IN_FLIGHT];
+	core::smart_refctd_ptr<video::IGPUFence> frameFences[FRAMES_IN_FLIGHT];
+	for (uint32_t i = 0u; i < FRAMES_IN_FLIGHT; ++i)
+	{
+		acquireSemaphores[i] = device->createSemaphore();
+		releaseSemaphores[i] = device->createSemaphore();
+		frameFences[i] = device->createFence(video::IGPUFence::E_CREATE_FLAGS::ECF_SIGNALED_BIT);
+	}
+
+	// Record commands in commandBuffers here
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		video::IGPUCommandBuffer::SImageMemoryBarrier undefToComputeTransitionBarrier;
+		undefToComputeTransitionBarrier.barrier.srcAccessMask = asset::EAF_TRANSFER_READ_BIT;
+		undefToComputeTransitionBarrier.barrier.dstAccessMask = asset::EAF_SHADER_READ_BIT;
```

Comment:
you want shader write bit

Reply:
unless you care about reading back content of old swapchain

Reply:
> you want shader write bit

Yes, because it is possible that some other invocation of the compute shader tries to write to the swapchain image before the layout transition (from UNDEFINED to GENERAL) actually occurs. So we want to make sure we hold off all the compute shader writes.

Reply:
Resolved